File: problems/0004-median-of-two-sorted-arrays/solution.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        
        if (nums1.length > nums2.length) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int len1 = nums1.length;
        int len2 = nums2.length;

        int left = 0;
        int right = len1;

        while (left <= right) {

            int x = (left + right) / 2;
            int y = (len1 + len2 + 1) / 2 - x;

            int left1  = (x == 0)    ? Integer.MIN_VALUE : nums1[x - 1];
            int right1 = (x == len1) ? Integer.MAX_VALUE : nums1[x];

            int left2  = (y == 0)    ? Integer.MIN_VALUE : nums2[y - 1];
            int right2 = (y == len2) ? Integer.MAX_VALUE : nums2[y];

            if (left1 <= right2 && left2 <= right1) {
                if ((len1 + len2) % 2 == 0) {
                    return (Math.max(left1, left2) + Math.min(right1, right2)) / 2.0;
                } else {
                    return Math.max(left1, left2);
                }

            } else if (left1 > right2) {
                right = x - 1;
            } else {
                left = x + 1;
            }
        }

        throw new IllegalArgumentException();
    }
}

File: problems/0004-median-of-two-sorted-arrays/solution.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        if len(nums1) > len(nums2):
            return self.findMedianSortedArrays(nums2, nums1)
        
        len1 = len(nums1)
        len2 = len(nums2)

        left = 0
        right = len1

        while left <= right:
            
            x = (left + right) // 2
            y = (len1 + len2 + 1) // 2 - x

            left1  = float('-inf') if (x == 0)    else nums1[x - 1]
            right1 = float('+inf') if (x == len1) else nums1[x]
            left2  = float('-inf') if (y == 0)    else nums2[y - 1]
            right2 = float('+inf') if (y == len2) else nums2[y]
        
            if left1 <= right2 and left2 <= right1:  
                if (len1 + len2) % 2 == 0:
                    return (max(left1, left2) + min(right1, right2)) / 2.0
                else:
                    return max(left1, left2)

            elif left1 > right2:
                right = x - 1
            else:
                left = x + 1